src/index/MmapBasedArray.h

```diff
@@ -0,0 +1,190 @@
+// Copyright 2018, University of Freiburg,
+// Chair of Algorithms and Data Structures.
+// Author: Johannes Kalmbach (joka921) <johannes.kalmbach@gmail.com>
+
+#pragma once
+
+#include <string>
+
+#include "../util/Exception.h"
+#include <exception>
+
+using std::string;
+
+// _________________________________________________________________________
+class UnititializedArrayException : public std::exception {
+  const char * what() const throw() {
+    return "Tried to access a DiskBasedArray which was closed or uninitialized\n";
+  }
+};
+
+// ________________________________________________________________
+struct VecInfo {
+  size_t _capacity;
+  size_t _bytesize;
+};
+
+// STL-like class which implements a dynamic array (similar to std::vector)
+// whose contents are stored persistently in a file on memory and are accessed
+// using memory mapping
+template<class T>
+class MmapBasedArray {
+ public:
+  virtual ~MmapBasedArray();
+  // TODO: other type aliases for convenience
+  using const_iterator = const T*;
+  using iterator = T*;
+
+  // __________________________________________________________________
+  size_t size() const {return _size;}
+
+  // standard iterator functions, each in a const and non-const version 
+  // begin
+  T* begin() { return _ptr;}
+  const T* begin() const {return _ptr;}
+  // end
+  T* end() { return _ptr + _size;}
+  const T* end() const { return _ptr + _size;}
+  // cbegin and cend
+  const T* cbegin() const { return _ptr;}
+  const T* cend() const { return _ptr + _size;}
+
+  // Element access
+  // without bounds checking
+  T& operator[](size_t idx) { throwIfUninitialized(); return _ptr[idx];}
+  const T& operator[](size_t idx) const { throwIfUninitialized(); return *(_ptr + idx);}
+
+  // with bounds checking
+  T& at(size_t idx) {
+    throwIfUninitialized();
+    if (idx >= _size) {
+      throw std::out_of_range();
+    }
+    return *(_ptr + idx);
+  }
+
+  // _____________________________________________________________
+  const T& at(size_t idx) const {
+    throwIfUninitialized();
+    if (idx >= _size) {
+      throw std::out_of_range();
+    }
+    return *(_ptr + idx);
+    AD_CHECK(idx < _size);
+  }
+
+  // we will never hold less than this capacity
+  static constexpr size_t MinCapacity = 100;
+
+  // Default constructor. This array is invalid until initialized by one of the
+  // setup methods;
+
+  // create Array of given size  fill with default value
+  // file contents will be overriden if existing!
+  // allows read and write access
+  void setupNewFile(size_t size, const T& defaultValue, string filename);
```

Comment:
`setupNewFile()` sounds like `open()` to me

Reply:
In this case I agree, I would refactor it using an argument of type std::ios_base::openmode and mimic the read/write semantics used with ifstreams.

Reply:
I decided to use two "tag"-classes. You will see if you like it better that way.